### PR TITLE
Commercial bundle dev server

### PR DESCRIPTION
**Pull request by chrislomaxjones**
## What does this change?

This PR introduces a new way to serve our commercial bundle, independent of the rest of Frontend. It can be divided into a few broad steps.

### 1. Setup a development server

Install [`webpack-dev-server`](https://webpack.js.org/configuration/dev-server/) (and `webpack-cli` to run command) as a development dependency.

We then add the necessary configuration properties to `webpack.config.commercial.dev.js` (the entrypoint for our development webpack configuration) in order to serve our standalone bundle on a specified port (I've chosen 3031 semi-arbitrarily, DCR uses 3030). This webpack configuration targets our standalone bundle as an entrypoint, so will ensure all of the files we care about are chunked appropriately.

This allows us to retrieve the commercial bundle from the local URL http://localhost:3031/graun.standalone.commercial.js. We can then pass this directly to DCR via an environment variable (since https://github.com/guardian/dotcom-rendering/pull/5572), circumventing Frontend in development entirely!

### 2. Fix dynamic imports

Unfortunately just doing the above is not enough to get dynamic imports working. This is because:
- In production we serve static assets from https://assets.guim.co.uk (which is backed by Amazon S3).
- In order to ensure our standalone bundle knows where to find bundles generated as dynamic imports, we have to override the `__webpack_public__path__` variable to point to https://assets.guim.co.uk in prod.
- Prior to this PR the value for `__webpack_public__path__` was constructed from values passed in on `window.guardian.page.config`.

When running a webpack dev server locally, we want to instead override the `__webpack_public_path__` to point to our local wepback dev server. We can do this easily enough, but it's necessary to inject some config from webpack into the bundle. We do this by passing an `OVERRIDE_BUNDLE` environment variable into our webpack config and use this to inject an environment variable _into_ our bundle. Using `OVERRIDE_BUNDLE` ensures we don't inject the override port into the bundle when we're running against Frontend.

## Does this change need to be reproduced in dotcom-rendering ?

- [ ] No
- [X] Yes - [This PR](https://github.com/guardian/dotcom-rendering/pull/5572) added the ability to override the bundle URL from the DCR side.

## What is the value of this and can you measure success?

This is another small step towards decoupling our Commercial code from the platform on which it will run. By adding this, in the future we should be able to more flexibly target and inject our bundle into different platforms (and onto platforms running in different environments).

Setting up a development server for our standalone code opens up the ability to run our commercial bundle against just DCR without spinning up the Frontend Scala Play applications. This has benefits for local testing - if you want to make a change to commercial code constrained to content types rendered by DCR. It will also will allow us to perform end-to-end testing against DCR in the future as well.

### Tested

- [X] Locally
- [x] On CODE

### CODE REVIEW

File: static/src/javascripts/bootstraps/standalone.commercial.ts [hunk 1]
```diff
@@ -46,9 +46,20 @@ import type { Modules } from './types';
 
 const { isDotcomRendering, page } = window.guardian.config;
 
-const assetsPath = page.frontendAssetsFullURL ?? page.assetsPath;
+const decideAssetsPath = () => {
+	const assetsPathPort: number | undefined = process.env.OVERRIDE_BUNDLE_PORT
+		? Number(process.env.OVERRIDE_BUNDLE_PORT)
```
Comment (domlander) [inline static/src/javascripts/bootstraps/standalone.commercial.ts:51]: Why do we convert this to a number?

Comment (chrislomaxjones) [reply static/src/javascripts/bootstraps/standalone.commercial.ts:51]: Reading this back, there's probably little point actually. My thinking was that a HTTP port is a number, but if we're just appending it to a string this might be unnecessary...

File: static/src/javascripts/bootstraps/standalone.commercial.ts [hunk 2]
```diff
@@ -46,9 +46,20 @@ import type { Modules } from './types';
 
 const { isDotcomRendering, page } = window.guardian.config;
 
-const assetsPath = page.frontendAssetsFullURL ?? page.assetsPath;
+const decideAssetsPath = () => {
```
Comment (domlander) [inline static/src/javascripts/bootstraps/standalone.commercial.ts:49]: I don't think we need lines 51 and 52. Also if we test whether `assetsPathPort` is falsy, then if the env variable were to be an empty string for some reason, then it will run the code in the `else` clause.
```suggestion
const decideAssetsPath = () => {
const assetsPathPort = process.env.OVERRIDE_BUNDLE_PORT

if (assetsPathPort) {
		return `http://localhost:${assetsPathPort}/`;
	} else {
		const assetsPath = page.frontendAssetsFullURL ?? page.assetsPath;
		return `${assetsPath}javascripts/commercial/`;
	}
}
```

Comment (domlander) [reply static/src/javascripts/bootstraps/standalone.commercial.ts:49]: I tried to do a Github code suggestion but I can't work it out yet 😅

Comment (chrislomaxjones) [reply static/src/javascripts/bootstraps/standalone.commercial.ts:49]: 🤦 Oh yeah, I should've just removed the ternary when we stopped casting to a Number, thanks for spotting! Thanks for the other point as well, makes sense!

File: static/src/javascripts/bootstraps/standalone.commercial.ts [hunk 3]
```diff
@@ -46,9 +46,18 @@ import type { Modules } from './types';
...
+	const assetsPathPort = process.env.OVERRIDE_BUNDLE_PORT;
+
+	if (assetsPathPort) {
+		return `http://localhost:${assetsPathPort}/`;
+	} else {
+		const assetsPath = page.frontendAssetsFullURL ?? page.assetsPath;
+		return `${assetsPath}javascripts/commercial/`;
+	}
```
Comment (domlander) [inline static/src/javascripts/bootstraps/standalone.commercial.ts:57]: Sorry to keep drip-feeding suggestions, but I was thinking we could make it clearer for the future reader that we are choosing to override the bundle in the `if` clause? Something like:
```
if (process.env.OVERRIDE_BUNDLE) {
    const assetsPathPort = process.env.OVERRIDE_BUNDLE_PORT;
	if (assetsPathPort) {
		return `http://localhost:${assetsPathPort}/`;
	}
}

const assetsPath = page.frontendAssetsFullURL ?? page.assetsPath;
return `${assetsPath}javascripts/commercial/`;
```
Feel free to ignore if you think the existing way is clearer!

**Review by domlander — APPROVED** [review]
Superb!

File: static/src/javascripts/bootstraps/standalone.commercial.ts (hunk 3 above)
Comment (chrislomaxjones) [reply static/src/javascripts/bootstraps/standalone.commercial.ts:57]: Good point! I actually think the first if will narrow `OVERRIDE_BUNDLE_PORT` from `string | undefined` to `string`, so we could avoid the `if (assetsPathPort) { ... }`

File: webpack.config.commercial.dev.js [hunk 4]
```diff
@@ -1,11 +1,30 @@
...
 	devtool: 'inline-source-map',
 	mode: 'development',
 	output: {
 		filename: `graun.standalone.commercial.js`,
 		chunkFilename: `graun.[name].commercial.js`,
 	},
+	plugins: shouldOverrideBundle
+		? [
+				new webpack.DefinePlugin({
+					'process.env.OVERRIDE_BUNDLE_PORT': JSON.stringify(port),
```
Comment (arelra) [inline webpack.config.commercial.dev.js:20]: Just wondering if this could be renamed and extended to:

- `OVERRIDE_ASSET_PATH` which can then be the value of the full path i.e. `http://localhost:3031/`

**Review by arelra — APPROVED** [review]
This is going to be super useful! 🚀

File: webpack.config.commercial.dev.js (hunk 4 above)
Comment (chrislomaxjones) [reply webpack.config.commercial.dev.js:20]: I see no reason why not 👍
